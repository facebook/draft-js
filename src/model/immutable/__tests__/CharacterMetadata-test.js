/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 */

'use strict';

jest.autoMockOff();

const CharacterMetadata = require('CharacterMetadata');
const {
  BOLD,
  BOLD_ITALIC,
  NONE,
  UNDERLINE,
} = require('SampleDraftInlineStyle');

describe('CharacterMetadata', () => {
  it('must have appropriate default values', () => {
    const character = CharacterMetadata.create();
    expect(character.getStyle().size).toBe(0);
    expect(character.getEntity()).toBe(null);
  });

  describe('Style handling', () => {
    const plain = CharacterMetadata.create();
    const bold = CharacterMetadata.create({style: BOLD});
    const fancy = CharacterMetadata.create({style: BOLD_ITALIC});

    it('must run `hasStyle` correctly', () => {
      expect(plain.hasStyle('BOLD')).toBe(false);
      expect(bold.hasStyle('BOLD')).toBe(true);
      expect(fancy.hasStyle('BOLD')).toBe(true);
      expect(plain.hasStyle('ITALIC')).toBe(false);
      expect(bold.hasStyle('ITALIC')).toBe(false);
      expect(fancy.hasStyle('ITALIC')).toBe(true);
    });

    it('must apply style', () => {
      const newlyBold = CharacterMetadata.applyStyle(plain, 'BOLD');
      expect(newlyBold.hasStyle('BOLD')).toBe(true);
      const alsoItalic = CharacterMetadata.applyStyle(newlyBold, 'ITALIC');
      expect(alsoItalic.hasStyle('BOLD')).toBe(true);
      expect(alsoItalic.hasStyle('ITALIC')).toBe(true);
    });

    it('must remove style', () => {
      const justBold = CharacterMetadata.removeStyle(fancy, 'ITALIC');
      expect(justBold.hasStyle('BOLD')).toBe(true);
      expect(justBold.hasStyle('ITALIC')).toBe(false);
      const justPlain = CharacterMetadata.removeStyle(justBold, 'BOLD');
      expect(justPlain.hasStyle('BOLD')).toBe(false);
      expect(justPlain.hasStyle('ITALIC')).toBe(false);
    });
  });

  describe('Entity handling', () => {
    const withoutEntity = CharacterMetadata.create();
    const withEntity = CharacterMetadata.create({entity: 'a'});

    it('must apply entity correctly', () => {
      const newKey = 'x';
      const modifiedA = CharacterMetadata.applyEntity(withoutEntity, newKey);
      const modifiedB = CharacterMetadata.applyEntity(withEntity, newKey);
      expect(modifiedA.getEntity()).toBe(newKey);
      expect(modifiedB.getEntity()).toBe(newKey);
    });

    it('must remove entity correctly', () => {
      const modifiedA = CharacterMetadata.applyEntity(withoutEntity, null);
      const modifiedB = CharacterMetadata.applyEntity(withEntity, null);
      expect(modifiedA.getEntity()).toBe(null);
      expect(modifiedB.getEntity()).toBe(null);
    });
  });

  describe('Metadata Reuse', () => {
    const empty = CharacterMetadata.create();
    const withStyle = CharacterMetadata.create({style: BOLD});
    const withTwoStyles = CharacterMetadata.create({style: BOLD_ITALIC});
    const withEntity = CharacterMetadata.create({entity: '1234'});
    const withStyleAndEntity = CharacterMetadata.create({
      entity: '1234',
      style: BOLD,
    });

    it('must reuse the same objects', () => {
      expect(CharacterMetadata.create()).toBe(empty);
      expect(CharacterMetadata.create({style: BOLD})).toBe(withStyle);
      expect(
        CharacterMetadata.create({style: BOLD_ITALIC})
      ).toBe(
        withTwoStyles
      );
      expect(CharacterMetadata.create({entity: '1234'})).toBe(withEntity);
      expect(
        CharacterMetadata.create({entity: '1234', style: BOLD})
      ).toBe(
        withStyleAndEntity
      );
    });

    it('must reuse objects by defaulting config properties', () => {
      expect(
        CharacterMetadata.create({style: BOLD, entity: null})
      ).toBe(
        withStyle
      );
      expect(
        CharacterMetadata.create({style: NONE, entity: '1234'})
      ).toBe(
        withEntity
      );

      const underlined = CharacterMetadata.create({
        style: UNDERLINE,
        entity: null,
      });

      expect(
        CharacterMetadata.create({style: UNDERLINE})
      ).toBe(
        underlined
      );
    });
  });
});
