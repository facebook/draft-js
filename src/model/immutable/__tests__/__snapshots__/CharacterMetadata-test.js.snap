// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply entity correctly 1`] = `
Immutable.OrderedSet [
  "x",
]
`;

exports[`must apply entity correctly 2`] = `
Immutable.OrderedSet [
  "a",
  "x",
]
`;

exports[`must apply style 1`] = `true`;

exports[`must apply style 2`] = `true`;

exports[`must apply style 3`] = `true`;

exports[`must have appropriate default values 1`] = `
Object {
  "entity": Array [],
  "style": Array [],
}
`;

exports[`must have appropriate default values 2`] = `0`;

exports[`must have appropriate default values 3`] = `Immutable.OrderedSet []`;

exports[`must remove entity correctly 1`] = `Immutable.OrderedSet []`;

exports[`must remove entity correctly 2`] = `
Immutable.OrderedSet [
  "a",
]
`;

exports[`must remove style 1`] = `true`;

exports[`must remove style 2`] = `false`;

exports[`must remove style 3`] = `false`;

exports[`must remove style 4`] = `false`;

exports[`must reuse objects by defaulting config properties 1`] = `true`;

exports[`must reuse objects by defaulting config properties 2`] = `true`;

exports[`must reuse objects by defaulting config properties 3`] = `true`;

exports[`must reuse the same objects 1`] = `true`;

exports[`must reuse the same objects 2`] = `true`;

exports[`must reuse the same objects 3`] = `true`;

exports[`must reuse the same objects 4`] = `true`;

exports[`must reuse the same objects 5`] = `true`;

exports[`must run \`hasStyle\` correctly 1`] = `false`;

exports[`must run \`hasStyle\` correctly 2`] = `true`;

exports[`must run \`hasStyle\` correctly 3`] = `true`;

exports[`must run \`hasStyle\` correctly 4`] = `false`;

exports[`must run \`hasStyle\` correctly 5`] = `false`;

exports[`must run \`hasStyle\` correctly 6`] = `true`;
